CppExport/src/visitors/Visitor.h

```diff
@@ -0,0 +1,131 @@
+
+struct VisitorData
+{
+		QList<ExportError> errors_;
+};
+
+class Visitor
+{
+	public:
+		Visitor();
+		Visitor(std::shared_ptr<VisitorData> data);
+
+		QList<ExportError> errors() const;
+
+	protected:
+		std::shared_ptr<VisitorData> data();
+
+		void required(Model::Node* parent, Model::Node* node, const QString& childName);
+		void notAllowed(Model::Node* node);
+		void notAllowed(Model::List* list);
+
+		void error(const QString& errorMessage);
+		void error(Model::Node* node, const QString& errorMessage);
+
+		template<class ListElement, class VisitorClass>
+		Export::SourceFragment* list(Model::TypedList<ListElement>* list, VisitorClass* v,
+											  const QString& fragmentType = QString());
+		template<class ListElement, class VisitorClass>
+		Export::SourceFragment* list(Model::TypedList<ListElement>* list, VisitorClass&& v,
+											  const QString& fragmentType = QString());
+
+		template<class ListElement, class VisitorClass, typename Predicate>
+		Export::SourceFragment* listWhere(Model::TypedList<ListElement>* list, VisitorClass* v,
+											  Predicate predicate, const QString& fragmentType = QString(),
+												bool* found = nullptr);
+		template<class ListElement, class VisitorClass, typename Predicate>
+		Export::SourceFragment* listWhere(Model::TypedList<ListElement>* list, VisitorClass&& v,
+											  Predicate predicate, const QString& fragmentType = QString(),
+												bool* found = nullptr);
+
+		template <class NodeType> Export::SourceFragment* declaration(NodeType* node);
+		template <class NodeType> Export::SourceFragment* statement(NodeType* node);
+		template <class NodeType> Export::SourceFragment* expression(NodeType* node);
+		template <class NodeType> Export::SourceFragment* element(NodeType* node);
+
+	private:
+		std::shared_ptr<VisitorData> data_;
+};
+
+inline std::shared_ptr<VisitorData> Visitor::data() { return data_; }
+
+inline QList<ExportError> Visitor::errors() const { return data_->errors_; }
+
+inline void Visitor::error(const QString& errorMessage) { data_->errors_.append(ExportError(errorMessage)); }
+inline void Visitor::error(Model::Node* node, const QString& errorMessage)
+{ data_->errors_.append(ExportError(node, errorMessage)); }
+
+template<class ListElement, class VisitorClass>
+Export::SourceFragment* Visitor::list(Model::TypedList<ListElement>* list, VisitorClass* v, const QString& fragmentType)
+{
+	auto fragment = new Export::CompositeFragment(list, fragmentType);
+	for (auto node : *list) *fragment << v->visit(node);
+	return fragment;
+}
+
+template<class ListElement, class VisitorClass>
+inline Export::SourceFragment* Visitor::list(Model::TypedList<ListElement>* list, VisitorClass&& v,
+												  const QString& fragmentType)
+{
+	return Visitor::list(list, &v, fragmentType);
+}
+
+template<class ListElement, class VisitorClass, typename Predicate>
+Export::SourceFragment* Visitor::listWhere(Model::TypedList<ListElement>* list, VisitorClass* v,
```

Comment:
Soo, we need changes here:
- Please do not introduce two additional functions, just extend the existing ones with one extra argument:
- Do not use `found`, check the returned fragment instead.
- Rename `predicate` to `filter`
- the names of the final methods should remain `list` and the predicate should appear as the last argument with a default value, so that existing cases work as before.